blog/2023-02-16-do-you-need-to-build-auth/index.md

```diff
@@ -0,0 +1,124 @@
+
+When I worked for Airbnb, there was a whole team dedicated to optimizing the sign-in experience for multiple platforms. They conducted numerous A/B tests to determine which combination had the highest conversion rate.
+
+It’s not practical to do so when a business is getting started. But we still need to try our best to make every piece right. Which platforms would you like to run the bookstore? You may start with mobile, web, or both. The exact design will depend on the authentication methods you chose:
+
+- Username and password: The easiest one, just several input boxes and buttons.
+- Passwordless: Enter phone / email, then send and verify a dynamic code.
+- Social sign-in: Read the docs from the chosen social identity provider, follow the visual guideline, handle the redirect logic, and finally link the social identity with the bookstore identity.
+
+More things to consider to make it better:
+
+- Do you need to combine the sign-in and registration process for a specific method?
+- Do you need a “forgot password” flow to allow customers to reset their passwords independently?
+
+If you opt for native development, the workload will almost double for one additional platform.
+
+### Security and token exchange
+
+Security can be a hidden iceberg. It’ll be great if you are familiar with OpenID Connect or OAuth 2.0, since they are widely used and battle-tested. OpenID Connect is relatively new but is designed for “user authentication”, which is a great fit for an online bookstore.
+
+Without delving into the details of the standards, there are still some things to consider:
+
+- Which encryption method should be used for passwords?
+- What is the process for standard authentication and authorization?
+- How does token exchange work (Access Token, Refresh Token, ID Token, etc.)?
+- How can signing keys be used in tokens and how can the signature be validated to protect resources?
+- How can client and server attacks be prevented?
+
+Finally, you can land a good sign-in experience and deliver it to your customers.
+
+### Authorization model
+
+As a bookstore, you need to separate resources for customers and sellers. For example, customers can browse all books, while sellers can manage their on-sale books. It's OK to start with simple if-else checks; however, as the business grows, you may need to leverage a more flexible model such as Role-based Access Control (RBAC) or Attribute-based Access Control (ABAC).
+
+I also wrote an article [CIAM 102: Authorization & Role-based Access Control](../2023-02-05-ciam-102-authz-and-rbac/index.md) to demonstrate basic authorization concepts and the RBAC model.
+
+## Make the decision
+
+You can see auth is not an “all or nothing” problem, since it involves multiple technical components and you or your team may have different tech expertise on these areas. It's important to break it down into smaller parts to gain a better understanding of the status quo.
```

Comment:
```suggestion
You can see auth is not an “all or nothing” problem, since it involves multiple technical components and you or your team may have different tech expertise in these areas. It's important to break it down into smaller parts to gain a better understanding of the status quo.
```